compiler/rustc_trait_selection/src/solve/eval_ctxt.rs

```diff
@@ -513,4 +538,87 @@ impl<'tcx> EvalCtxt<'_, 'tcx> {
+        self.infcx.create_next_universe()
+    }
+
+    pub(super) fn instantiate_canonical_var(
+        &self,
+        cv_info: CanonicalVarInfo<'tcx>,
+        universe_map: impl Fn(ty::UniverseIndex) -> ty::UniverseIndex,
+    ) -> ty::GenericArg<'tcx> {
+        self.infcx.instantiate_canonical_var(DUMMY_SP, cv_info, universe_map)
+    }
+
+    pub(super) fn translate_substs(
+        &self,
+        param_env: ty::ParamEnv<'tcx>,
+        source_impl: DefId,
+        source_substs: ty::SubstsRef<'tcx>,
+        target_node: specialization_graph::Node,
+    ) -> ty::SubstsRef<'tcx> {
+        crate::traits::translate_substs(
+            self.infcx,
+            param_env,
+            source_impl,
+            source_substs,
+            target_node,
+        )
+    }
+
+    pub(super) fn register_ty_outlives(&self, ty: Ty<'tcx>, lt: ty::Region<'tcx>) {
+        self.infcx.register_region_obligation_with_cause(ty, lt, &ObligationCause::dummy());
+    }
+
+    pub(super) fn register_region_outlives(&self, a: ty::Region<'tcx>, b: ty::Region<'tcx>) {
+        // `b : a` ==> `a <= b`
+        // (inlined from `InferCtxt::region_outlives_predicate`)
+        self.infcx.sub_regions(
+            rustc_infer::infer::SubregionOrigin::RelateRegionParamBound(DUMMY_SP),
+            b,
+            a,
+        );
+    }
+
+    /// Computes the list of goals required for `arg` to be well-formed
+    pub(super) fn well_formed_goals(
+        &self,
+        param_env: ty::ParamEnv<'tcx>,
+        arg: ty::GenericArg<'tcx>,
+    ) -> Option<impl Iterator<Item = Goal<'tcx, ty::Predicate<'tcx>>>> {
+        crate::traits::wf::unnormalized_obligations(self.infcx, param_env, arg)
+            .map(|obligations| obligations.into_iter().map(|obligation| obligation.into()))
+    }
+
+    #[instrument(level = "debug", skip(self), ret)]
+    pub(super) fn compute_external_query_constraints(
+        &self,
+    ) -> Result<ExternalConstraints<'tcx>, NoSolution> {
+        // Cannot use `take_registered_region_obligations` as we may compute the response
+        // inside of a `probe` whenever we have multiple choices inside of the solver.
+        let region_obligations = self.infcx.inner.borrow().region_obligations().to_owned();
+        let region_constraints = self.infcx.with_region_constraints(|region_constraints| {
+            make_query_region_constraints(
+                self.tcx(),
+                region_obligations
+                    .iter()
+                    .map(|r_o| (r_o.sup_type, r_o.sub_region, r_o.origin.to_constraint_category())),
+                region_constraints,
+            )
+        });
+        let opaque_types = self.infcx.clone_opaque_types_for_query_response();
+        Ok(self
+            .tcx()
+            .mk_external_constraints(ExternalConstraintsData { region_constraints, opaque_types }))
+    }
+
+    pub(super) fn canonicalize<T: TypeFoldable<TyCtxt<'tcx>>>(
```

Comment:
I'm not super satisfied that this fn has to live here... but I couldn't think of anything better, and modifying the `Canonicalizer` to take an `EvalCtxt` requires re-exporting a bunch of different var probing/root var resolving helper fns in  `EvalCtxt` that I didn't like the look of.

----

Alternatively, we could move `rustc_trait_selection::solve::canonical` to `rustc_trait_selection::solve::eval_ctxt::canonical` -- then all the `EvalCtxt` functions are within the right scope to access `self.infcx` even if it's private.

I'd probably then want to move `rustc_trait_selection::solve::canonical::canonicalizer` to somewhere new.

Reply:
I dislike having canonicalization (except for `evaluate_added_goals_and_make_query_response`) methods on `EvalCtxt`. These are also kinda incorrect in the public api, similar to the `infcx` itself.

you could use `pub(in super::canonicalize)` or alternative make the canonicalization module a submodule.

Reply:
what :o

why does this not work?

```rust
mod parent {
    mod this {
        pub(in super::sibling) fn foo() {}
    }
    
    mod sibling {
        fn bar() {
            super::this::foo();
        }
    }
}
```

Reply:
I don't think it was ever intended to be able to make things public in sister -- or arbitrary -- modules.